File: scala/bank-account/src/main/scala/BankAccount.scala
import BankAccount._

class BankAccount {

  private var balance: Balance = Some(0)

  def getBalance: Balance = balance

  def incrementBalance(amount: Money): Balance = {
    synchronized {
      balance = balance map (_ + amount)
    }
    balance
  }

  def closeAccount(): Unit =
    balance = None
}

object BankAccount {
  def apply() = new BankAccount

  type Money = Int
  type Balance = Option[Money]
}

File: haskell/bank-account/BankAccount.hs
module BankAccount ( BankAccount, openAccount, closeAccount, getBalance, incrementBalance ) where

import Data.IORef

type Account = Maybe Integer
type BankAccount = IORef Account

openAccount :: IO BankAccount
openAccount = newIORef $ Just 0

closeAccount :: BankAccount -> IO ()
closeAccount = flip atomicWriteIORef Nothing

getBalance :: BankAccount -> IO Account
getBalance = readIORef

incrementBalance :: BankAccount -> Integer -> IO Account
incrementBalance bankAccount amount =
  atomicModifyIORef bankAccount incrAccount
  where
    incrAccount = pair . fmap (+ amount)
    pair x = (x,x)
